README
                                  INTRODUCTION

FLASH (Fast Length Adjustment of SHort reads) is an accurate and fast tool
to merge paired-end reads that were generated from DNA fragments whose
lengths are shorter than twice the length of reads.  Merged read pairs result
in unpaired longer reads, which are generally more desired in genome
assembly and genome analysis processes.

Briefly, the FLASH algorithm considers all possible overlaps at or above a
minimum length between the reads in a pair and chooses the overlap that
results in the lowest mismatch density (proportion of mismatched bases in
the overlapped region).  Ties between multiple overlaps are broken by
considering quality scores at mismatch sites.  When building the merged
sequence, FLASH computes a consensus sequence in the overlapped region.
More details can be found in the original publication
(http://bioinformatics.oxfordjournals.org/content/27/21/2957.full).

Limitations of FLASH include:
   - FLASH cannot merge paired-end reads that do not overlap.
   - FLASH is not designed for data that has a significant amount of indel
     errors (such as Sanger sequencing data).  It is best suited for Illumina
     data.

                                  INSTALLATION

On UNIX-compatible systems, including GNU/Linux and Mac OS X, you must compile
FLASH from source.  The only dependency, other than functions that are expected
to be available in the C library, is the zlib data compression library.  To
install FLASH, download the tarball, untar it, and compile the code using the
provided Makefile:

    $ tar xzf FLASH-1.2.11.tar.gz
    $ cd FLASH-1.2.11
    $ make

The executable file that is produced is named 'flash'.  To run it from the
command line you must copy it to a location on your $PATH variable, or else run
it with a path including a directory, such as "./flash".

FLASH also runs on Windows, and you can compile it on Windows using MinGW.
However, for convenience you may instead download a standalone Windows binary
from the SourceForge page (https://sourceforge.net/projects/flashpage/).

                                     USAGE

Please compile FLASH and run `flash --help' to see command-line usage
information and information about input/output files.

                                 MULTITHREADING

By default, FLASH uses multiple threads.  There are "combiner" threads that do
the actual read combining, as well as up to 5 threads that are used for I/O (up
to 2 readers, up to 3 writers).  The default number of combiner threads is the
number of processors; however, it can be adjusted with the -t  option (long
option: --threads).

When multiple combiner threads are used, the order of the combined and
uncombined reads in the output files will be nondeterministic.  If you need to
enforce that the output reads appear in the same order as the input, you must
specify --threads=1.

                                  PERFORMANCE

Since the FLASH algorithm considers each read pair independently, FLASH will, by
default, process read pairs in parallel.  FLASH v1.2.9 and later also make use
of vector instructions available on modern x86 CPUs.  Consequently, FLASH works
quite fast, even with low-cost computing resources.  As an example, we ran FLASH
v1.2.9 on a laptop with a dual-core 2.3 GHz AMD x86_64 processor and it
processed one million 101-bp read pairs in 11.6 seconds with the default
parameters.  Less than 2 MB of memory was used.  Actual timing results will
vary, but they will depend primarily on the number of CPUs available, the speed
of each CPU, and on the I/O speed of reading the input files and writing the
output files.  FLASH is designed to be scalable to dozens of processors,
although its speed may be limited by I/O in such cases.

                                   ACCURACY

With reads' error rate of 1% or less, FLASH processes over 99% of read pairs
correctly.  With error rate of 2%, FLASH processes over 98% of read pairs
correctly when default parameters are used. With more aggressive parameters
(i.e., -x 0.35), FLASH processes over 90% of read pairs correctly even when the
error rate is 5%.

                                  PUBLICATION

Title:   FLASH: fast length adjustment of short reads to improve genome assemblies
Authors: Tanja Magoč and Steven L. Salzberg
URL:     http://bioinformatics.oxfordjournals.org/content/27/21/2957.full

                                    LICENSE

FLASH is released under the GNU General Public License Version 3 or later (see
COPYING).

			         MODIFICATIONS

Modications that were undertook:
	-Outies logic was changed
		Over lap was still Allgined like this. . .
			<--------------                   Read 1
				 -------------->          Read 2
		But This was not kept, the overlapping area was kept
				 ------ Read 1
				 ------ Read 2

	-"Engulf" Case was added
		-If read 2 (or read 1) was engulfed by the other - they would not overlap
		-This allows for that to be an overlap - along with a new argument min-overlap-outie
	-"Threshold Quality" was added
		-There was some significant problems with primer dimers - this removed most of them
		-This is one of the first steps for flashdbcamp.
			
                          COMMENTS/QUESTIONS/REQUESTS

Send an e-mail to streett.david@gmail.com for information on the modified flash2

Original unmodified version is available from the SourceForge page:

https://sourceforge.net/projects/flashpage/


NEWS
Version 1.2.11
--------------
- Fixed a bug that caused FLASH to crash under the following set of
  circumstances: the --allow-outies option (added in v1.2.10) was specified; and
  in a read pair, read 1 was longer than read 2; and the chosen merge aligned
  the end of read 2 to the very beginning of read 1, e.g.

	  Read 1:  ----------------------------------->
	  Read 2:  <-------------------------

  This bug was extremely likely to crash the program.  Therefore, the output of
  successful runs of FLASH should not have been affected.

Version 1.2.10
--------------
- FLASH now supports combining read pairs in outie orientation.  See help for
  the new --allow-outies option.
- Added warning messages in a few cases where it may be helpful to specify
  different parameters.

Version 1.2.9
-------------------
- The core algorithm of FLASH has been optimized using SSE and SSE2
  instructions.  On x86_64 CPUs it now can run over twice as fast.
- The help output of FLASH has been improved.
- Added supported for tab-delimited input and output.
- FLASH now displays more informative error messages in some cases.
- Reorganized some of the code so that it's easier to modify.
- Reading the data from standard input now works correctly.
- With --to-stdout specified, informational messages are now written to standard
  error rather than omitted.

Version 1.2.8
-------------------
- FLASH's default behavior when assigning quality scores to mismatch sites in
  the overlap region of combined reads has been improved.  See the documentation
  for the new --cap-mismatch-quals option (which can be provided to request the
  old behavior if desired) for more information about this change.

Version 1.2.7
-------------------
- FLASH is now supported on Windows.  See the README.
- Minor bug-fixes regarding printing of error messages.
- Fixed a bug where the time that FLASH took to run may have been mis-calculated
  on 32-bit systems in an unlikely case.  This only affected the printed
  informational output.
- Updated PERFORMANCE section in README to be accurate for the current version
  of FLASH.

Version 1.2.6
-------------------
- FLASH now works on files larger than 2GiB on 32-bit operating systems, unless
  the system zlib library does not support such files.  64-bit operating systems
  are unaffected as they already supported large files by default.

Version 1.2.5
-------------------
- FLASH will now compile and run on Mac OS X.

Version 1.2.4
-------------------
- Support for piping the output of FLASH through arbitrary compression programs
  has been added via the --compress-prog option.  The --compress-prog-args
  and/or --output-suffix options also may be useful.
- FLASH with the --interleaved option will now use an interleaved output format
  in addition to assuming interleaved input.  Use --interleaved-input or
  --interleaved-output to specify only interleaved input or only interleaved
  output, respectively.
- The entirely single-threaded configuration of FLASH is no longer supported.
  You can still use --threads=1 to use only 1 combiner thread, however.

Version 1.2.3
-------------------
- Support for interleaved reads (-I, --interleaved), which may be read from one
  FASTQ file or piped in from stdin
- Fix --read-len option
- Fix default --fragment-len-stddev option
- Fix reading FASTQ files with embedded zero bytes
- Fix memory leaks
- Added license file

Version 1.2.2
-------------------
- Fix linking problem in Makefile

Version 1.2.1
-------------------
- There is no longer any read length or tag length limit, although you may still
specify the --read-len parameter for the purpose of automatically computing the
maximum overlap.
- FLASH now prints some informational output while it's running, unless running
with --quiet or --to-stdout.

Version 1.2
-------------------

- Support for writing extended fragments directly to standard output.
- Support for writing compressed output files.
- Support for multithreading.  There are multiple I/O threads as well as
multiple combiner threads.  Speedup of 3.6 times has been observed.  Compiling
without support for multithreading is still supported.

Version 1.1
-------------------
Rewrote the entire program.
- Removed at least 3 instances of undefined behavior (such as using
uninitialized variables).
- Performance improvements. Neglecting I/O, the program now runs over twice as
fast.  This improvement is not due to a change in algorithm; I expect it is due
mainly to the rewritten align_position() function.
- Use getopt_long() for option parsing.  Now, long options are accepted, and
options do not have to be given after the mate pair FASTQ files.
- Removed many fixed-size buffers.  This saves memory, removes many buffer
overflow problems, and removes the READ_LENGTH parameter.  Note: the maximum tag
length is still hard-coded as 1024 bytes.
- Much more error checking, so the program will cleanly fail if something goes
wrong.
- Use a Makefile instead of a shell script.
- The '.histogram' file is written a bit more narrowly (80 columns), but other
than that, the output of the program should be the same (unless it was affected
by any of the cases of undefined behavior I found).
- Some changes to the MANUAL, although I didn't feel like changing the
information that is redundant with the program help.
- Note: I am calling this version 1.1, but feel free to call it something else.
The version number is also listed in the version() function in flash.c.

Version v1.0.3
--------------
1. FLASH allows reads files to be in gzip format.
2. FLASH builds a histogram of fragment lengths produced by merged read pairs.

combine_reads.h
#ifndef _FLASH_COMBINE_READS_H_
#define _FLASH_COMBINE_READS_H_

#include <stdbool.h>

struct read;

/* Parameters for the core algorithm of FLASH.  See the help output for more
 * information.  */
struct combine_params {
	/* --min-overlap  */
	int min_overlap;


	/* --min-overlap-outie */
	int min_overlap_outie;

	/* --max-overlap  */
	int max_overlap;

	/* --max-mismatch-density  */
	float max_mismatch_density;

	/* --cap-mismatch-quals  */
	bool cap_mismatch_quals;

	/* --allow-outies  */
	bool allow_outies;

	bool discard_reads;
	int percent_cutoff;
	int qual_score_cutoff;
	
};

/* Result of a call to combine_reads()  */
enum combine_status {
	/* The reads could not be combined.  */
	NOT_COMBINED = 0,

	JUNK_READ = -5,
	/* The reads were combined in "innie" orientation, like the following:
	 *
	 * ---------->
	 *     <------------
	 *
	 * (Note: read_2 is reverse complemented before the call to
	 * combine_reads()).  */
	COMBINED_AS_INNIE,

	/* The reads were combined in "outie" orientation, like the following:
	 *
	 * <----------
	 *     ------------>
	 *
	 * (Note: read_2 is reverse complemented before the call to
	 * combine_reads()).  */
	COMBINED_AS_OUTIE,
};

extern enum combine_status
combine_reads(const struct read *read_1, const struct read *read_2,
	      struct read *combined_read,
	      const struct combine_params *params);

#endif /* _FLASH_COMBINE_READS_H_  */


combine_reads.c
/*
 * combine_reads.c:  This file contains the code implementing the core algorithm
 * to combine reads in FLASH.
 */

/*
 * Copyright (C) 2012 Tanja Magoc
 * Copyright (C) 2012, 2013, 2014 Eric Biggers
 *
 * This file is part of FLASH, a fast tool to merge overlapping paired-end
 * reads.
 *
 * FLASH is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 3 of the License, or (at your option)
 * any later version.
 *
 * FLASH is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License
 * along with FLASH; if not, see http://www.gnu.org/licenses/.
 */

#include "combine_reads.h"
#include "read.h"
#include "util.h"

#include <assert.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>

#if defined(__GNUC__) && defined(__SSE2__)
#  define WITH_SSE2
#endif

#ifdef WITH_SSE2
#  include <xmmintrin.h>
#  include <emmintrin.h>
#endif

#ifdef WITH_SSE2

/* Sum the values an 8 x 8 bit vector and return a 32-bit result.  */
static inline uint32_t
hsum32_v8(__m128i v)
{
	v = _mm_sad_epu8(v, _mm_set1_epi8(0));
	return (uint32_t)_mm_extract_epi16(v, 0) +
	       (uint32_t)_mm_extract_epi16(v, 4);
}

/* Sum the values an 8 x 16 bit vector and return a 32-bit result.  */
static inline uint32_t
hsum32_v16(__m128i v)
{
	__m128i mask = _mm_set1_epi32(0x0000ffff);
	v = _mm_add_epi32(v & mask, _mm_srli_si128(v, 2) & mask);
	v = _mm_add_epi32(v, _mm_srli_si128(v, 4));
	v = _mm_add_epi32(v, _mm_srli_si128(v, 8));
	return _mm_cvtsi128_si32(v);
}

#endif /* WITH_SSE2 */

/*
 * Compute mismatch statistics between two sequences.
 *
 * @seq_1, @seq_2:
 *	The two sequences to compare (ASCII characters A, C, G, T, N).
 * @qual_1, @qual_2:
 *	Quality scores for the two sequences, based at 0.
 * @haveN
 *	As an optimization, this can be set to %false to indicate that neither
 *	sequence contains an uncalled base (represented as an N character).
 * @len_p
 *	Pointer to the length of the sequence.  This value will be updated to
 *	subtract the number of positions at which an uncalled base (N) exists in
 *	either sequence.
 * @num_mismatches_ret
 *	Location into which to return the number of positions at which the bases
 *	were mismatched.
 * @mismatch_qual_total_ret
 *	Location into which to return the sum of lesser quality scores at
 *	mismatch sites.
 */

static inline void
compute_mismatch_stats(const char * restrict seq_1,
		       const char * restrict seq_2,
		       const char * restrict qual_1,
		       const char * restrict qual_2,
		       bool haveN,
		       int * restrict len_p,
		       unsigned * restrict num_mismatches_ret,
		       unsigned * restrict mismatch_qual_total_ret)
{
	int num_uncalled = 0;
	unsigned num_mismatches = 0;
	unsigned mismatch_qual_total = 0;
	int len = *len_p;

	if (haveN) {
		for (int i = 0; i < len; i++) {
			if (seq_1[i] == 'N' || seq_2[i] == 'N') {
				num_uncalled++;
			} else {
				if (seq_1[i] != seq_2[i])  {
					num_mismatches++;
					mismatch_qual_total += min(qual_1[i], qual_2[i]);
				}
			}
		}
	} else {
		/* This part of the 'if' statement is for optimization purposes
		 * only; its behavior is equivalent to the block above, except
		 * this block assumes there are no N characters in the input,
		 * and therefore no further checks for N's are needed.
		 *
		 * Note: this optimization is only useful if most reads don't
		 * contain N characters.  */

	#ifdef WITH_SSE2

		/* Optional vectorized implementation (about twice as fast as
		 * nonvectorized on x86_64).  */

		while (len >= 16) {

			/* 16 x 8 bit counters for number of mismatches  */
			__m128i num_mismatches_v8 = _mm_set1_epi8(0);

			/* 8 x 16 bit counters for mismatch quality total  */
			__m128i mismatch_qual_total_v16 = _mm_set1_epi16(0);

			/* The counters of num_mismatches_v8 will overflow if
			 * 256 mismatches are detected at the same position
			 * modulo 16 bytes.  So, don't process 4096 or more
			 * bytes before reducing the counters.
			 *
			 * mismatch_qual_total_v16 would overflow even faster,
			 * but we use 16-bit counters for it.  */

			int todo = min(len, 255 * 16) & ~0xf;
			len -= todo;

			do {
				/* Load 16 bases  */
				__m128i s1_v8 = _mm_loadu_si128((const void *)seq_1);
				__m128i s2_v8 = _mm_loadu_si128((const void *)seq_2);

				/* Load 16 quality scores  */
				__m128i q1_v8 = _mm_loadu_si128((const void *)qual_1);
				__m128i q2_v8 = _mm_loadu_si128((const void *)qual_2);

				/* Compare bases with each other and negate the
				 * result.  This will produce 0xff in bytes
				 * where the bases differ and 0x00 in bytes
				 * where the bases were the same.  */
				__m128i cmpresult = ~_mm_cmpeq_epi8(s1_v8, s2_v8);

				/* Tally mismatched bases.  Subtracting 0x00 and
				 * 0xff is equivalent to adding 0 and 1,
				 * respectively.  */
				num_mismatches_v8 = _mm_sub_epi8(num_mismatches_v8,
								 cmpresult);

				/* Tally quality scores for mismatched bases.
				 */

				/* Get minimum of each quality score.  */
				__m128i qmin_v8 = _mm_min_epu8(q1_v8, q2_v8);

				/* Select only quality scores at mismatch sites
				 */
				__m128i qadd_v8 = qmin_v8 & cmpresult;

				/* Double the precision (8 => 16 bits) and tally  */
				__m128i qadd_v16_1 = _mm_unpacklo_epi8(qadd_v8,
								       _mm_set1_epi8(0));

				__m128i qadd_v16_2 = _mm_unpackhi_epi8(qadd_v8,
								       _mm_set1_epi8(0));

				mismatch_qual_total_v16 = _mm_add_epi16(mismatch_qual_total_v16,
									qadd_v16_1);

				mismatch_qual_total_v16 = _mm_add_epi16(mismatch_qual_total_v16,
									qadd_v16_2);

				/* Advance pointers  */
				seq_1 += 16, seq_2 += 16;
				qual_1 += 16, qual_2 += 16;
				todo -= 16;
			} while (todo);

			/* Reduce the counters.  */
			num_mismatches += hsum32_v8(num_mismatches_v8);
			mismatch_qual_total += hsum32_v16(mismatch_qual_total_v16);
		}

	#endif /* WITH_SSE2  */

	#if 0
		/* Verify the values computed by the vectorized implementation.
		 */
		{
			int veclen = *len_p & ~0xf;
			const char *_seq_1 = seq_1 - veclen;
			const char *_seq_2 = seq_2 - veclen;
			const char *_qual_1 = qual_1 - veclen;
			const char *_qual_2 = qual_2 - veclen;
			unsigned _num_mismatches = 0;
			unsigned _mismatch_qual_total = 0;
			for (int i = 0; i < veclen; i++) {
				if (_seq_1[i] != _seq_2[i])  {
					_num_mismatches++;
					_mismatch_qual_total += min(_qual_1[i], _qual_2[i]);
				}
			}
			assert(num_mismatches == _num_mismatches);
			assert(mismatch_qual_total == _mismatch_qual_total);
		}
	#endif

		/* Process any remainder that wasn't processed by the vectorized
		 * implementation.  */
		for (int i = 0; i < len; i++) {
			if (seq_1[i] != seq_2[i])  {
				num_mismatches++;
				mismatch_qual_total += min(qual_1[i], qual_2[i]);
			}
		}
	}

	/* Return results in pointer arguments  */
	*num_mismatches_ret = num_mismatches;
	*mismatch_qual_total_ret = mismatch_qual_total;
	*len_p -= num_uncalled;
}

#define NO_ALIGNMENT INT_MIN



static inline int *
pair_align(const struct read *read_1, const struct read *read_2,
	   int min_overlap, int max_overlap, int min_overlap_outie, float max_mismatch_density,
	   bool allow_outies, bool * was_outie)
{
	bool haveN = memchr(read_1->seq, 'N', read_1->seq_len) ||
		     memchr(read_2->seq, 'N', read_2->seq_len);

	/* Best (smallest) mismatch density that has been found so far in an
	 * overlap. */
	float best_mismatch_density = max_mismatch_density + 1.0f;
	float best_qual_score = 0.0f;
	int best_position = NO_ALIGNMENT;
	bool best_was_outie = false;
	bool doing_outie = false;
	int start;
	int end;
	int read_offset = 0;
	int best_offset = 0;
again:
	/* Require at least min_overlap bases overlap, and require that the
	 * second read is not overlapped such that it is completely contained in
	 * the first read.  */
	start = max(0, read_1->seq_len - read_2->seq_len);
	end = read_1->seq_len - min_overlap + 1;

	if (start == 0) {
		read_offset = (read_2->seq_len - read_1->seq_len);

		if (read_offset < 0) {
			read_offset *= -1;
		}
	}



	for (int i = start; i < end; i++) {
		unsigned num_mismatches;
		unsigned mismatch_qual_total;
		int overlap_len = read_1->seq_len - i;

		/*This modification allows for engulf cases of the read*/
		compute_mismatch_stats(read_1->seq + i,
				       read_2->seq + read_offset,
				       read_1->qual + i,
				       read_2->qual + read_offset,
				       haveN,
				       &overlap_len,
				       &num_mismatches,
				       &mismatch_qual_total);

		/*Logic to make sure read makes minimum requirements*/
		if (((!doing_outie && overlap_len >= min_overlap) || (doing_outie && overlap_len >= min_overlap_outie))) {
			float score_len = (float)min(overlap_len, max_overlap);
			float qual_score = mismatch_qual_total / score_len;
			float mismatch_density = num_mismatches / score_len;
				
			if (mismatch_density <= best_mismatch_density &&
			    (mismatch_density < best_mismatch_density ||
			     qual_score < best_qual_score))
			{
				best_qual_score       = qual_score;
				best_mismatch_density = mismatch_density;
				best_position         = i;
				best_offset           = read_offset;
				best_was_outie        = doing_outie;
			}
		}
		if (read_offset != 0) {
			read_offset -= (i + 1);
			i--;
		}
		
	}

	/*At one point I will remove the goto statments - in non-modified code*/
	if (allow_outies) {
		const struct read *tmp = read_1;
		read_1 = read_2;
		read_2 = tmp;
		allow_outies = false;
		doing_outie = true;
		goto again;
	}

	int *position_and_offset = (int*)malloc(2 * sizeof(int));

	position_and_offset[0] = best_position;
	position_and_offset[1] = best_offset;


	if (best_mismatch_density > max_mismatch_density) {
		position_and_offset[0] = NO_ALIGNMENT;
		position_and_offset[1] = NO_ALIGNMENT;
	}
//		return position_and_offset;

	*was_outie = best_was_outie;
	return position_and_offset;
}

/* Fills in the combined read from the specified alignment.  */
static void
generate_combined_read(const struct read *read_1,
		       const struct read *read_2,
		       struct read *combined_read,
		       int overlap_begin,
		       int read_offset,
	               bool was_outie,
		       bool cap_mismatch_quals)
{
	/* Length of the overlapping part of two reads.  */

	int overlap_len = read_1->seq_len - overlap_begin;
	/* Length of the part of the second read not overlapped with the first
	 * read.  */
	int remaining_len = read_2->seq_len - overlap_len;

	int combined_seq_len = read_1->seq_len + remaining_len;

	if (was_outie) {
		//Case of outie, Engulf case
		if (read_offset != 0) {
			combined_seq_len = read_offset + read_1->seq_len;
		//Outie, remove both sides
		} else {
			combined_seq_len = overlap_len;
		}
	} else {
		//Innie, engulf case
		if (read_offset != 0) {
			combined_seq_len = read_1->seq_len + read_offset;	
		}
	}
	const char * restrict seq_1 = read_1->seq;
	const char * restrict seq_2 = read_2->seq;
	const char * restrict qual_1 = read_1->qual;
	const char * restrict qual_2 = read_2->qual;
	char * restrict combined_seq;
	char * restrict combined_qual;

	/*Allocates the correct size for each condition*/
	if (combined_read->seq_bufsz < combined_seq_len || was_outie || read_offset != 0) {
	
                combined_read->seq = xrealloc(combined_read->seq,
                                              combined_seq_len);
                combined_read->seq_bufsz = combined_seq_len;
	
	}



	if (combined_read->qual_bufsz < combined_seq_len || was_outie || read_offset != 0) {
		combined_read->qual = xrealloc(combined_read->qual,
					       combined_seq_len);
		combined_read->qual_bufsz = combined_seq_len;
	}

	/* David Modification
 	 * --------|------------
 	 *          ------------|---------
 	 *          Only keeps the overlying parts
 	 */

                combined_seq = combined_read->seq;
	                combined_qual = combined_read->qual;

                combined_read->seq_len = combined_seq_len;
                combined_read->qual_len = combined_seq_len;


	/* Copy the beginning of read 1 (not in the overlapped region).  */
	if (!was_outie && read_offset == 0) {
		while (overlap_begin--) {
			*combined_seq++ = *seq_1++;
			*combined_qual++ = *qual_1++;
		}
	} else if (read_offset != 0) {

		while (read_offset--) {
			*combined_seq++ = *seq_2++;
			*combined_qual++ = *qual_2++;
		}
		read_offset = -1;
		overlap_len = read_1->seq_len;

	}
	/* Copy the overlapped region.  */
	while (overlap_len-- > 0) {

		if (*seq_1 == *seq_2) {
			/* Same base in both reads.  Take the higher quality
			 * value. */
			*combined_seq = *seq_1;
			*combined_qual = max(*qual_1, *qual_2);
		} else {
			/* Different bases in the two reads; use the higher
			 * quality one.
			 *
			 * The old way of calculating the resulting quality
			 * value (params->cap_mismatch_quals == %true) is to use
			 * the lower quality value, and use a quality value of
			 * at most 2 (+ phred_offset in the final output--- here
			 * the quality values are all scaled to start at 0).
			 * The motivation for this behavior is that the read
			 * combination shows there was sequencing error at the
			 * mismatch location, so the corresponding base call in
			 * the combined read should be given a low quality
			 * score.
			 *
			 * The new way (params->cap_mismatch_quals == %false,
			 * default as of FLASH v1.2.8) is to use the absolute
			 * value of the difference in quality scores, but at
			 * least 2.  This allows a base call with a high quality
			 * score to override a base call with a low quality
			 * score without too much penalty.
			 */

			if (cap_mismatch_quals)
				*combined_qual = min(min(*qual_1, *qual_2), 2);
			else
				*combined_qual = max(abs(*qual_1 - *qual_2), 2);

			if (*qual_1 > *qual_2) {
				*combined_seq = *seq_1;
			} else if (*qual_1 < *qual_2) {
				*combined_seq = *seq_2;
			} else {
				/* Same quality value; take the base from the
				 * first read if the base from the second read
				 * is an 'N'; otherwise take the base from the
				 * second read. */

				if (*seq_2 == 'N')
					*combined_seq = *seq_1;
				else
					*combined_seq = *seq_2;
			}
		}
		combined_seq++;
		combined_qual++;
		seq_1++;
		seq_2++;
		qual_1++;
		qual_2++;
	}
	/* Copy the end of read 2 (not in the overlapped region).  */
	if (!was_outie && read_offset == 0 && remaining_len > 0) {
		while (remaining_len--) {
			*combined_seq++ = *seq_2++;
			*combined_qual++ = *qual_2++;
		}
	/*This check and sees if there is an offset, but also that the offset wasn't for the innie case*/
	}/* else if (read_offset > 0) {
			while (read_offset--) {
				*combined_seq++ = *seq_2++;
			}
	}*/
}

/* This is the entry point for the core algorithm of FLASH.  The following
 * function attempts to combine @read_1 with @read_2, and writes the result into
 * @combined_read.  COMBINED_AS_INNIE or COMBINED_AS_OUTIE is returned if
 * combination was successful.  COMBINED_AS_OUTIE is only possible if
 * params->allow_outies is set.
 *
 * Note: @read_2 is provided to this function after having been
 * reverse-complemented.  Hence, the code just aligns the reads in the forward
 * orientation, which is equivalent to aligning the original reads in the
 * desired reverse-complement orientation.
 *
 * Please see the help output of FLASH for the description of the min_overlap,
 * max_overlap, and max_mismatch_density parameters.  (--min-overlap,
 * --max-overlap, and --max-mismatch-density on the command line).
 *
 * You may also want to read the original FLASH publication for a description of
 * the algorithm used here:
 *
 *  Title:   FLASH: fast length adjustment of short reads to improve genome assemblies
 *  Authors: Tanja Magoč and Steven L. Salzberg
 *  URL:     http://bioinformatics.oxfordjournals.org/content/27/21/2957.full
 *
 */






/*Function for Junk - removal of primer dimers
 *50 percent of quality scores must be greater the 2 as default*/

double check_for_junk_read(char * qual_1, char * qual_2, int cut_off) {

	double count = 0;
	double junk = 0;
	int i_1 = 0, i_2 = 0;


	/*Could vectorize this in the future*/
	while (qual_1[i_1] != '\0' && qual_2[i_2] != '\0') {
		if (qual_1[i_1] != '\0') {
			if (qual_1[i_1]  <= cut_off) {
				junk++;
			}
			count++;
			i_1++;
		}
	
		if (qual_2[i_2] != '\0') {
			if (qual_2[i_2] <= cut_off) {
				junk++;
			}
			count++;
			i_2++;
		}
	}
	/*Both doubles so do not have to worry about interger division*/
	return junk/count;
}


enum combine_status
combine_reads(const struct read *read_1, const struct read *read_2,
	      struct read *combined_read,
	      const struct combine_params *params)
{
	int overlap_begin, read_offset, *overlap_and_offset;
	enum combine_status status;
	bool was_outie = false;

	/*check read isn't junk;*/
	if (params->discard_reads) {
		double percent_junk = check_for_junk_read(read_1->qual, read_2->qual, params->qual_score_cutoff);
		if (percent_junk*100 > params->percent_cutoff) {
			return JUNK_READ;
		}	
	}
	/* Do the alignment.  */

	
	
	overlap_and_offset = pair_align(read_1, read_2,
				   params->min_overlap,
				   params->max_overlap,
				   params->min_overlap_outie,
				   params->max_mismatch_density,
				   params->allow_outies,
				   &was_outie);
	overlap_begin = overlap_and_offset[0];
	read_offset = overlap_and_offset[1];

	/*
	 * If overlap_begin == NO_ALIGNMENT, then no sufficient overlap between
	 * the reads was found.
	 *
	 * If !@was_outie, then the pair forms an "innie" overlap, and
	 * overlap_begin is the 0-based position in read_1 at which read_2
	 * begins.  (Shown below with read 2 already reverse complemented!)
	 *
	 *		0	  overlap_begin
	 *	        |         |
	 *	Read 1: ------------------>
	 *	Read 2:           ---------------------->
	 *
	 * If @was_outie, then the pair forms an "outie" overlap, and
	 * overlap_begin is the 0-based position in read_2 at which read_1
	 * begins. (Shown below with read 2 already reverse complemented!)
	 *
	 *	        0         overlap_begin
	 *	        |         |
	 *	Read 2: ------------------>
	 *	Read 1:           ---------------------->
	 */

	if (overlap_begin == NO_ALIGNMENT)
		return NOT_COMBINED;

	if (!was_outie) {
		status = COMBINED_AS_INNIE;
	} else {
		const struct read *tmp;

		/* Simplify generation of the combined read by turning the outie
		 * case into the innie case.  */

		tmp = read_1;
		read_1 = read_2;
		read_2 = tmp;

		status = COMBINED_AS_OUTIE;
		/*
		 * Now it's just:
		 *
		 *		0	  overlap_begin
		 *	        |         |
		 *	Read 1: ------------------>
		 *	Read 2:           ---------------------->
		 *
		 * The same as the "innie" case.
		 */
	}

	/* Fill in the combined read.  */
	generate_combined_read(read_1, read_2, combined_read,
			       overlap_begin, read_offset, was_outie, params->cap_mismatch_quals);
	return status;
}
